projects/train/train/model/supervised.py

```diff
@@ -20,3 +20,67 @@ def train_step(self, batch: tuple[Tensor, Tensor]) -> Tensor:
 
     def score(self, X):
         return self(X)
+
+
+class SupervisedAframeS4(SupervisedAframe):
+    def __init__(self, arch: SupervisedArchitecture, *args, **kwargs) -> None:
+        super().__init__(arch, *args, **kwargs)
+
+    def forward(self, X):
+        X = X.transpose(-1, -2)
+        return self.model(X)
+
+    def configure_optimizers(self):
+        """
+        S4 requires a specific optimizer setup.
+
+        The S4 layer (A, B, C, dt) parameters typically
+        require a smaller learning rate (typically 0.001),
+        with no weight decay.
+
+        The rest of the model can be trained with a higher learning rate
+        (e.g. 0.004, 0.01) and weight decay (if desired).
+        """
+        if not torch.distributed.is_initialized():
+            world_size = 1
+        else:
+            world_size = torch.distributed.get_world_size()
+
+        # All parameters in the model
+        all_parameters = list(self.model.parameters())
+
+        # General parameters don't contain the special _optim key
+        params = [p for p in all_parameters if not hasattr(p, "_optim")]
+
+        # Create an optimizer with the general parameters
+        lr = self.hparams.learning_rate * world_size
+        self._logger.info(f"Scaled lr by {world_size} to {lr}")
+        optimizer = torch.optim.AdamW(
+            params, lr=lr, weight_decay=self.hparams.weight_decay
+        )
+
+        # Add parameters with special hyperparameters
+        hps = [
+            getattr(p, "_optim")
+            for p in all_parameters
+            if hasattr(p, "_optim")
+        ]
+        hps = [
+            dict(s)
+            for s in sorted(
+                list(dict.fromkeys(frozenset(hp.items()) for hp in hps))
+            )
+        ]  # Unique dicts
+        for hp in hps:
+            params = [
+                p for p in all_parameters if getattr(p, "_optim", None) == hp
+            ]
+            optimizer.add_param_group({"params": params, **hp})
+
+        # Create a lr scheduler
+        scheduler = torch.optim.lr_scheduler.CosineAnnealingLR(
+            optimizer, self.trainer.max_epochs
```

Comment:
I think you need to either set this to `self.trainer.estimated_stepping_batches`, or set the `interval` below to `epochs` - double check me on this though

Reply:
True, yeah, missed that